src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryUntypedObject.cs

```diff
@@ -297,37 +297,71 @@ public static FormulaValue Float_UO(EvalVisitor runner, EvalVisitorContext conte
             return GetTypeMismatchError(irContext, BuiltinFunctionsCore.Value_UO.Name, DType.Number.GetKindString(), impl);
         }
 
-        public static FormulaValue Text_UO(EvalVisitor runner, EvalVisitorContext context, IRContext irContext, UntypedObjectValue[] args)
-        {
-            var impl = args[0].Impl;
-
+        public static FormulaValue Text_UO(EvalVisitor runner, EvalVisitorContext context, IRContext irContext, FormulaValue[] args)
+        {
+            var arg0 = (UntypedObjectValue)args[0];
+            var impl = arg0.Impl;
+
+            FormulaValue newArg0 = null;
             if (impl.Type == FormulaType.String)
             {
-                var str = impl.GetString();
-                return new StringValue(irContext, str);
+                var str = impl.GetString();
+                var stringArg0 = new StringValue(irContext, str);
+                newArg0 = stringArg0;
+
+                // The second argument to Text is a format, if it exists then we need to convert the string to a type that
+                // can be formatted. We try to convert to number first and if that fails we try dates.
+                if (args.Length > 1)
+                {
+                    var numArg0 = Value_UO(runner, context, IRContext.NotInSource(FormulaType.Number), new UntypedObjectValue[] { arg0 });
+                    if (numArg0 is not ErrorValue)
+                    {
+                        newArg0 = numArg0;
+                    }
+                    else
+                    {
+                        // Use DateTimeValue_UO to apply format restrictions
+                        var dateArg0 = DateTimeValue_UO(runner, context, IRContext.NotInSource(FormulaType.DateTime), new UntypedObjectValue[] { arg0 });
+                        if (dateArg0 is not ErrorValue)
+                        {
+                            newArg0 = dateArg0;
+                        }
+                    }
+                }
             }
             else if (impl.Type is ExternalType et && et.Kind == ExternalTypeKind.UntypedNumber)
             {
                 var str = impl.GetUntypedNumber();
-                return new StringValue(irContext, str);
+                newArg0 = new StringValue(irContext, str);
+
+                // If a second argument is provided, then we cannot guarantee that the original precison will be
+                // preserved. We convert to a number, either Float or Decimal depending on the settings.

```

Comment:
Unfortunately, it won't be Float or Decimal, as written it will always be Float as FormulaType.Number was used to call Value_UO on line 341.  To this point, the interpreter hasn't needed to know if NumberIsFloat was being used and so it isn't passed in.  That's in general a good thing.  Value_UO figures it out because CheckTypes for Value_UO provides the return type through the irContext.  But there is no return type in this case (besides String for the Text function).  Is it time for the intepreter to know more directly?  Include it as ambient state in the irContext?  Thiking about it...  Some of our hosts, like PVA, are heavily Decimal, I don't think we should favor Float just because of this information gap.

Reply:
The argument is now pre-processed. Code is injected to either convert to float or decimal in IRTranslator.